src/puppetlabs/puppetdb/http/command.clj

```diff
@@ -36,12 +38,57 @@
 
         :else (app req)))))
 
+(defmacro with-chan
+  "Bind chan-sym to init-chan in the scope of the body, calling async/close! in
+  a finally block.
+
+  (with-chan [my-chan (clojure.core.async/chan)] ...)"
+  [[chan-sym init-chan] & body]
+  `(let [~chan-sym ~init-chan]
+     (try
+       (do ~@body)
+       (finally (async/close! ~chan-sym)))))
+
+(defmacro with-sub
+  "Subscribe chan to the given 'from' pub, in the scope of the body."
+  [{:keys [from topic chan]} & body]
+  `(let [p# ~from, t# ~topic, c# ~chan]
+     (try
+       (async/sub p# t# c#)
+       (do ~@body)
+       (finally
+         (async/unsub p# t# c#)))))
+
+(defn- blocking-submit-command
+  "Submit a command by calling do-submit-fn and block until it completes.
+  Subscribes to response-pub on the topic of the commands uuid, waiting up to
+  completion-timeout-ms."
+  [do-submit-fn response-pub uuid completion-timeout-ms]
+  (with-chan [response-chan (async/chan)]
+    (with-sub {:from response-pub :topic (str uuid) :chan response-chan}
+      (let [timeout-chan (async/timeout completion-timeout-ms)
```

Comment:
It doesn't; in fact it's slightly more work for the computer if we close it early.